rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -15,406 +15,238 @@
+    auto total_samples = rx_metadata.numberofreceivesamples() * total_antennas;
 
+    DEBUG_MSG("Total samples in data message: " << total_samples);
 
-int main(int argc, char **argv){
-    // REVIEW #35 main is > 200 lines, kind of large maybe the entire filter setup process could be moved out?
-    GOOGLE_PROTOBUF_VERIFY_VERSION; // REVIEW #4 state what this does? macro to verify headers and lib are same version.
+    dp->allocate_and_copy_rf_samples(total_samples);
+    dp->allocate_and_copy_first_stage_filters(filters.get_first_stage_bandpass_taps_h().data(),
+                                                filters.get_first_stage_bandpass_taps_h().size());
 
-    auto driver_options = DriverOptions(); // REVIEW #15 Should do a quick sanity check on all config options for driver, sig
-    auto sig_options = SignalProcessingOptions(); // #26 REVIEW Should the naming be updated along with DSP ?
-    auto rx_rate = driver_options.get_rx_rate(); // #5 REVIEW What units is rx_rate in?
-    zmq::context_t sig_proc_context(1); // REVIEW #4 - what is "1"?
+    auto num_output_samples_1 = rx_freqs.size() *
+                                  rx_metadata.numberofreceivesamples()/first_stage_dm_rate *
+                                  total_antennas;
 
-    zmq::socket_t driver_socket(sig_proc_context, ZMQ_PAIR);
-    driver_socket.bind("ipc:///tmp/feeds/1"); // REVIEW #29 Should this be in a config file? Sort of a magic string right now
+    dp->allocate_first_stage_output(num_output_samples_1);
 
-    // REVIEW #1 Need a comment here to explain which blocks these 3 sockets talk to, the driver socket is obvious, but these three may not be
-    zmq::socket_t radarctrl_socket(sig_proc_context, ZMQ_PAIR); // REVIEW #26 Name of radarctrl may need to be updated to be consistent with our discussion on Friday March 10th
-    radarctrl_socket.bind("ipc:///tmp/feeds/2");
+    gpuErrchk(cudaStreamAddCallback(dp->get_cuda_stream(),
+                  DSPCore::initial_memcpy_callback, dp, 0));
 
-    zmq::socket_t ack_socket(sig_proc_context, ZMQ_PAIR);
-    ack_socket.bind("ipc:///tmp/feeds/3"); // REVIEW #37 check the errno on all binds
+    call_decimate<DecimationType::bandpass>(dp->get_rf_samples_p(),
+      dp->get_first_stage_output_p(),
+      dp->get_first_stage_bp_filters_p(), first_stage_dm_rate,
+      rx_metadata.numberofreceivesamples(), filters.get_first_stage_lowpass_taps().size(), rx_freqs.size(),
+      total_antennas, "First stage of decimation", dp->get_cuda_stream());
 
-    zmq::socket_t timing_socket(sig_proc_context, ZMQ_PAIR);
-    timing_socket.bind("ipc:///tmp/feeds/4");
 
-    auto gpu_properties = get_gpu_properties();
-    print_gpu_properties(gpu_properties);
+    // When decimating, we go from one set of samples for each antenna in the first stage
+    // to multiple sets of reduced samples for each frequency in further stages. Output samples are
+    // grouped by frequency with all samples for each antenna following each other
```

Comment:
As we discussed in your office, the poster was slightly misleading and these comments/code are correct